File: forms/outpatient-outcome-report/front-end-form-with-svelte/src/lib/engine/clinical-domain.ts
import type { AssessmentData, DomainGrade, FiredRule } from './types';

/**
 * Grade the Clinical domain from the clinician-rated outcome classification.
 *
 * A = Resolved
 * B = Improved
 * C = Unchanged
 * D = Worsened
 * E = Died
 * '' = No classification entered
 */
export function gradeClinical(data: AssessmentData): { grade: DomainGrade; rules: FiredRule[] } {
	const cls = data.clinicalOutcome.outcomeClassification;
	const rules: FiredRule[] = [];

	let grade: DomainGrade = '';

	switch (cls) {
		case 'resolved':
			grade = 'A';
			rules.push({
				id: 'CLIN-001',
				domain: 'Clinical',
				description: 'Outcome classified as Resolved',
				grade: 'A'
			});
			break;
		case 'improved':
			grade = 'B';
			rules.push({
				id: 'CLIN-002',
				domain: 'Clinical',
				description: 'Outcome classified as Improved',
				grade: 'B'
			});
			break;
		case 'unchanged':
			grade = 'C';
			rules.push({
				id: 'CLIN-003',
				domain: 'Clinical',
				description: 'Outcome classified as Unchanged',
				grade: 'C'
			});
			break;
		case 'worsened':
			grade = 'D';
			rules.push({
				id: 'CLIN-004',
				domain: 'Clinical',
				description: 'Outcome classified as Worsened',
				grade: 'D'
			});
			break;
		case 'died':
			grade = 'E';
			rules.push({
				id: 'CLIN-005',
				domain: 'Clinical',
				description: 'Outcome classified as Died',
				grade: 'E'
			});
			break;
		default:
			// No classification — insufficient data
			grade = '';
			break;
	}

	return { grade, rules };
}

File: forms/outpatient-outcome/back-end-with-loco/src/engine/clinical_domain.rs
//! Clinical domain module.

use super::types::{AssessmentData, DomainGrade, FiredRule};

/// Clinical result.
pub struct ClinicalResult {
    /// Grade.
    pub grade: DomainGrade,
    /// Rules.
    pub rules: Vec<FiredRule>,
}

/// Grade the Clinical domain from the clinician-rated outcome classification.
///
/// A = Resolved
/// B = Improved
/// C = Unchanged
/// D = Worsened
/// E = Died
/// ''  = No classification entered
pub fn grade_clinical(data: &AssessmentData) -> ClinicalResult {
    let cls = data.clinical_outcome.outcome_classification.as_str();
    let mut rules: Vec<FiredRule> = Vec::new();
    let grade: DomainGrade = match cls {
        "resolved" => {
            rules.push(FiredRule {
                id: "CLIN-001".to_string(),
                domain: "Clinical".to_string(),
                description: "Outcome classified as Resolved".to_string(),
                grade: "A".to_string(),
            });
            "A".to_string()
        }
        "improved" => {
            rules.push(FiredRule {
                id: "CLIN-002".to_string(),
                domain: "Clinical".to_string(),
                description: "Outcome classified as Improved".to_string(),
                grade: "B".to_string(),
            });
            "B".to_string()
        }
        "unchanged" => {
            rules.push(FiredRule {
                id: "CLIN-003".to_string(),
                domain: "Clinical".to_string(),
                description: "Outcome classified as Unchanged".to_string(),
                grade: "C".to_string(),
            });
            "C".to_string()
        }
        "worsened" => {
            rules.push(FiredRule {
                id: "CLIN-004".to_string(),
                domain: "Clinical".to_string(),
                description: "Outcome classified as Worsened".to_string(),
                grade: "D".to_string(),
            });
            "D".to_string()
        }
        "died" => {
            rules.push(FiredRule {
                id: "CLIN-005".to_string(),
                domain: "Clinical".to_string(),
                description: "Outcome classified as Died".to_string(),
                grade: "E".to_string(),
            });
            "E".to_string()
        }
        _ => String::new(),
    };
    ClinicalResult { grade, rules }
}
